include/revng/ADT/GenericGraph.h

```diff
@@ -353,11 +370,557 @@ class BidirectionalNode
+    std::swap(*SuccessorIt, Predecessor->Successors.back());
+    Predecessor->Successors.pop_back();
+
+    std::swap(*Iterator, Predecessors.back());
+    Predecessors.pop_back();
+
+    return Iterator;
+  }
+
+public:
+  auto removeSuccessor(ConstSuccessorEdgeIterator Iterator) {
+    return SuccessorEdgeIterator(removeSuccessorImpl(Iterator.getCurrent()),
+                                 SuccessorFilters::ToView);
+  }
+  auto removeSuccessor(ConstSuccessorIterator Iterator) {
+    return SuccessorIterator(removeSuccessorImpl(Iterator.getCurrent()),
+                             SuccessorFilters::ToNeighbor);
+  }
+  auto removeSuccessor(DerivedType const *S) {
+    return SuccessorIterator(removeSuccessorImpl(findSuccessorImpl(S)),
+                             SuccessorFilters::ToNeighbor);
+  }
+
+  auto removePredecessor(ConstPredecessorEdgeIterator Iterator) {
+    return PredecessorEdgeIterator(removePredecessorImpl(Iterator.getCurrent()),
+                                   PredecessorFilters::ToView);
+  }
+  auto removePredecessor(ConstPredecessorIterator Iterator) {
+    return PredecessorIterator(removePredecessorImpl(Iterator.getCurrent()),
+                               PredecessorFilters::ToNeighbor);
+  }
+  auto removePredecessor(DerivedType const *S) {
+    return PredecessorIterator(removePredecessorImpl(findPredecessorImpl(S)),
+                               PredecessorFilters::ToNeighbor);
+  }
+
+public:
+  auto removeSuccessor(SuccessorEdgeIterator Iterator) {
+    return removeSuccessor(
+      ConstSuccessorEdgeIterator(Iterator.getCurrent(),
+                                 SuccessorFilters::ToConstView));
+  }
+  auto removeSuccessor(SuccessorIterator Iterator) {
+    return removeSuccessor(
+      ConstSuccessorIterator(Iterator.getCurrent(),
+                             SuccessorFilters::ToConstNeighbor));
+  }
+  auto removePredecessor(PredecessorEdgeIterator Iterator) {
+    return removePredecessor(
+      ConstPredecessorEdgeIterator(Iterator.getCurrent(),
+                                   PredecessorFilters::ToConstView));
+  }
+  auto removePredecessor(PredecessorIterator Iterator) {
+    return removePredecessor(
+      ConstPredecessorIterator(Iterator.getCurrent(),
+                               PredecessorFilters::ToConstNeighbor));
+  }
+
+public:
+  MutableEdgeNode &disconnect() {
+    for (auto It = Successors.begin(); It != Successors.end();)
+      It = removeSuccessorImpl(It);
+    for (auto It = Predecessors.begin(); It != Predecessors.end();)
+      It = removePredecessorImpl(It);
+    return *this;
+  }
+
+public:
+  SuccessorEdgeIterator successorEdgeAt(size_t Index) {
+    return SuccessorEdgeIterator(std::next(Successors.begin(), Index),
+                                 SuccessorFilters::ToView);
+  }
+  ConstSuccessorEdgeIterator successorEdgeAt(size_t Index) const {
+    return ConstSuccessorEdgeIterator(std::next(Successors.begin(), Index),
```

Comment:
For the same reason I prefer `++Counter` over `Counter++`. Theoretically, there's one less memory access (`std::next` doesn't need to modify the temporary iterator it got as an input (unlike `operator++()`).

In practice, compiler probably optimizes them to be the same thing anyway, so yeah, it's no more than a pure preference.

Wait, in this case it's different, because `Index` is passed to `std::next`, but I'll leave what I already wrote as the explanation for similar behavior elsewhere.

Reply:
Fair.